File: java/src/main/java/io/aurigraph/dlt/sdk/namespace/TransactionsApi.java
package io.aurigraph.dlt.sdk.namespace;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import io.aurigraph.dlt.sdk.AurigraphClient;
import io.aurigraph.dlt.sdk.dto.Transaction;
import io.aurigraph.dlt.sdk.dto.TransactionReceipt;
import io.aurigraph.dlt.sdk.dto.TransactionSubmitRequest;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Namespace for /api/v11/transactions endpoints.
 */
public class TransactionsApi {

    private final AurigraphClient client;

    public TransactionsApi(AurigraphClient client) {
        this.client = client;
    }

    public List<Transaction> recent(int limit) {
        JsonNode root = client.get("/transactions/recent?limit=" + limit, JsonNode.class);
        JsonNode list = root.isArray() ? root
                : (root.has("transactions") ? root.get("transactions")
                : (root.has("items") ? root.get("items") : null));
        if (list == null || !list.isArray()) return new ArrayList<>();
        try {
            return client.mapper().convertValue(list, new TypeReference<List<Transaction>>() {});
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    public TransactionReceipt submit(TransactionSubmitRequest tx) {
        return client.post("/transactions", tx, TransactionReceipt.class);
    }

    public Transaction get(String txHash) {
        String enc = URLEncoder.encode(txHash, StandardCharsets.UTF_8);
        return client.get("/transactions/" + enc, Transaction.class);
    }
}

File: android/aurigraph-sdk/src/main/kotlin/io/aurigraph/sdk/apis/TransactionsApi.kt
package io.aurigraph.sdk.apis

import io.aurigraph.sdk.AurigraphClient
import io.aurigraph.sdk.models.Transaction
import io.aurigraph.sdk.models.TransactionReceipt
import io.aurigraph.sdk.models.TransactionSubmitRequest
import kotlinx.serialization.json.JsonObject

/**
 * Transaction submission and querying.
 */
class TransactionsApi(private val client: AurigraphClient) {

    /** Submit a new transaction to the network. */
    suspend fun submit(request: TransactionSubmitRequest): TransactionReceipt =
        client.post("/transactions", request)

    /** Get a specific transaction by hash. */
    suspend fun get(txHash: String): Transaction = client.get("/transactions/$txHash")

    /** List recent transactions. */
    suspend fun listRecent(limit: Int = 20): JsonObject =
        client.get("/transactions/recent?limit=$limit")
}
